libraries/AP_AIS/AP_AIS.cpp

```diff
@@ -0,0 +1,764 @@
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Automatic Identification System, https://gpsd.gitlab.io/gpsd/AIVDM.html
+
+// ToDo: enable receiving of the Mavlink AIS message, type bitmask?
+
+#include "AP_AIS.h"
+
+#if HAL_AIS_ENABLED
+
+#include <AP_Logger/AP_Logger.h>
+#include <GCS_MAVLink/GCS_MAVLink.h>
+#include <GCS_MAVLink/GCS.h>
+
+#define LIST_SIZE_DEFAULT 25
+#define TIME_OUT_DEFAULT 600
+#define MAVLINK_RESEND_TIME_MS 30000 // only resend mavlink message for the same vehicle after this time if there has been no change
+
+const AP_Param::GroupInfo AP_AIS::var_info[] = {
+
+    // @Param: TYPE
+    // @DisplayName: AIS recever type
+    // @Description: AIS recever type
+    // @Values: 0:None,1:NMEA AIVDM message
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO_FLAGS("TYPE", 1, AP_AIS, _type, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: LIST_MAX
+    // @DisplayName: AIS vessel list size
+    // @Description: AIS list size of nearest vessels. Longer lists take longer to refresh with lower SRx_ADSB values.
+    // @Range: 1 100
+    // @User: Advanced
+    AP_GROUPINFO("LIST_MAX", 2, AP_AIS, _max_list, LIST_SIZE_DEFAULT),
+
+    // @Param: TIME_OUT
+    // @DisplayName: AIS vessel time out
+    // @Description: if no updates are received in this time a vessel will be removed from the list
+    // @Units: s
+    // @Range: 1 2000
+    // @User: Advanced
+    AP_GROUPINFO("TIME_OUT", 3, AP_AIS, _time_out, TIME_OUT_DEFAULT),
+
+    // @Param: OPTIONS
+    // @DisplayName: AIS options
+    // @Description: Bitmask of AIS options
+    // @Bitmask: 0:Log all AIVDM messages,1:Log only unsupported AIVDM messages,2:Log decoded messages
+    // @User: Advanced
+    AP_GROUPINFO("OPTIONS", 4, AP_AIS, _options, AIS_OPTIONS_LOG_UNSUPPORTED_RAW),
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_AIS::AP_AIS() :
+    _list(8)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+// Initialize the AIS object and prepare it for use
+void AP_AIS::init()
+{
+    if (!enabled()) {
+        return;
+    }
+
+    _uart = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_AIS, 0);
+    if (_uart == nullptr) {
+        return;
+    }
+
+    _uart->begin(AP::serialmanager().find_baudrate(AP_SerialManager::SerialProtocol_AIS, 0));
+}
+
+// update AIS, expected to be called at 20hz
+void AP_AIS::update()
+{
+    if (!_uart || !enabled()) {
+        return;
+    }
+
+    // read any available lines
+    u_int32_t nbytes = MAX(_uart->available(),1024U);
```

Comment:
```suggestion
    uint32_t nbytes = MAX(_uart->available(),1024U);
```

Second time we've seen these weird types :-)